### DISCUSSION TITLE
Get data from OUT Interrupt Endpoint

### DISCUSSION THREAD

**Discussion by mikeDonahoe** [sekigon-gonnoc/Pico-PIO-USB#68] [Q&A] [answered]
I might be missing something, but I seem to be having an issue receiving data from the host on PIO-USB configured as a device.
Is pio_usb_get_in_data intended to be used for receiving data from the host?  I was only able to get an ACK when I used

if(pio_usb_device_transfer(0x1,tempIn,64))

Which does grab data from the host, but it does not match what was sent.

![image](https://user-images.githubusercontent.com/6437324/215297877-b64a2d11-853b-4347-9d78-ed0effcd1ccc.png)
![image](https://user-images.githubusercontent.com/6437324/215303947-ea0400a5-0672-42aa-a02d-7135f5dda2c5.png)

I'm able to send data without issue. 

-Thanks

```

 bLength                :   0x12 (18 bytes)
 bDescriptorType        :    0x1 Device
 bcdUSB                 :  0x200 USB 2.0
 bDeviceClass           :    0x0 Specified at interface
 bDeviceSubClass        :    0x0
 bDeviceProtocol        :    0x0
 bMaxPacketSize0        :   0x40 (64 bytes)
 idVendor               : 0x7f17
 idProduct              : 0x0011
 bcdDevice              :  0x100 Device 1.0
 iManufacturer          :    0x1 BrickLabs
 iProduct               :    0x2 PiNucleo
 iSerialNumber          :    0x3 123456
 bNumConfigurations     :    0x1
  CONFIGURATION 1: 100 mA ==================================
   bLength              :    0x9 (9 bytes)
   bDescriptorType      :    0x2 Configuration
   wTotalLength         :   0x29 (41 bytes)
   bNumInterfaces       :    0x1
   bConfigurationValue  :    0x1
   iConfiguration       :    0x0 
   bmAttributes         :   0x80 Bus Powered
   bMaxPower            :   0x32 (100 mA)
    INTERFACE 0: Human Interface Device ====================
     bLength            :    0x9 (9 bytes)
     bDescriptorType    :    0x4 Interface
     bInterfaceNumber   :    0x0
     bAlternateSetting  :    0x0
     bNumEndpoints      :    0x2
     bInterfaceClass    :    0x3 Human Interface Device
     bInterfaceSubClass :    0x0
     bInterfaceProtocol :    0x0
     iInterface         :    0x0 
      ENDPOINT 0x81: Interrupt IN ==========================
       bLength          :    0x7 (7 bytes)
       bDescriptorType  :    0x5 Endpoint
       bEndpointAddress :   0x81 IN
       bmAttributes     :    0x3 Interrupt
       wMaxPacketSize   :   0x40 (64 bytes)
       bInterval        :    0xa
      ENDPOINT 0x1: Interrupt OUT ==========================
       bLength          :    0x7 (7 bytes)
       bDescriptorType  :    0x5 Endpoint
       bEndpointAddress :    0x1 OUT
       bmAttributes     :    0x3 Interrupt
       wMaxPacketSize   :   0x40 (64 bytes)
       bInterval        :    0xa
```

[1204307.389323] usb 1-1.4.4: new full-speed USB device number 109 using xhci_hcd
[1204312.814845] usb 1-1.4.4: New USB device found, idVendor=7f17, idProduct=0011, bcdDevice= 1.00
[1204312.814853] usb 1-1.4.4: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[1204312.814857] usb 1-1.4.4: Product: PiNucleo
[1204312.814859] usb 1-1.4.4: Manufacturer: BrickLabs
[1204312.814861] usb 1-1.4.4: SerialNumber: 123456
[1204312.918038] hid-generic 0003:7F17:0011.0094: hiddev4,hidraw4: USB HID v1.11 Device [BrickLabs PiNucleo] on usb-0000:63:00.1-1.4.4/input0

**Comment by mikeDonahoe** [comment, marked as answer]
Got VSCode set up so I could debug the library.  The following change seems to correct the issue with the first two bytes being wrong.  I tested pushing a 64byte payload with a non-zero value at the very end and I was able to receive it.  As of right now, I don't have a good understanding of how the usb_rx_buffer is filled in this scenario to know whether this is correct.  @sekigon-gonnoc thoughts?

![image](https://user-images.githubusercontent.com/6437324/215354489-0b918500-3eee-4617-9142-6fb319977b2d.png)

![image](https://user-images.githubusercontent.com/6437324/215354572-653e1c28-d555-4809-8c1c-1b6d7968a312.png)

```
diff --git a/src/pio_usb_device.c b/src/pio_usb_device.c
index 380951d..23ad402 100644
--- a/src/pio_usb_device.c
+++ b/src/pio_usb_device.c
@@ -172,7 +172,7 @@ static void __no_inline_not_in_flash_func(usb_device_packet_handler)(void) {
 
     if (ep->has_transfer) {
       if (res >= 0) {
-        memcpy(ep->app_buf, pp->usb_rx_buffer, res);
+        memcpy(ep->app_buf, pp->usb_rx_buffer + 2, res);
         pio_usb_ll_transfer_continue(ep, res);
       }
     }

```

**Reply by sekigon-gonnoc** [reply]
@mikeDonahoe It is correct. We have to discard first two byte(SYNC, DATA0/1) to get payloads. Could you submit PR?

**Reply by mikeDonahoe** [reply]
@sekigon-gonnoc Yes, I will do that after work.  Thanks
